From f7057774bf14a5a3730d1caa720266c78dcdfcd4 Mon Sep 17 00:00:00 2001
From: Alan Winkowski <walan@marvell.com>
Date: Mon, 30 Oct 2017 16:46:32 +0200
Subject: [PATCH 1247/1345] net: mvpp2x: fix double vlan parser detection

commit  fd82d4488ef6ee57b571fbafd33bcb70817c35aa from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- Before this patch, all traffic received with double VLAN
  set according to 802.1AD (outer vlan 0x88a8) was wrongly parsed
  as having single vlan by parser and no L3/L4 parsing was made.
- There was a mismatch between outer and inner VLAN in parser TCAM
- After this patch, parser TCAM is able to match outer VLAN
  set to either 0x8100 or 0x88a8 and inner VLAN set to 0x8100.
- Fixes A7K8K-3698

Change-Id: Ic3549ad740a5db30f7d1ea2b6d59fa54e2483155
Signed-off-by: Alan Winkowski <walan@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/45692
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/46858
Reviewed-by: Nir Erez <nerez@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c
index c998639..e80ab76 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c
@@ -1997,8 +1997,8 @@ static int mv_pp2x_prs_etype_init(struct mv_pp2x_hw *hw)
 
 /* Configure vlan entries and detect up to 2 successive VLAN tags.
  * Possible options:
- * 0x8100, 0x88A8
- * 0x8100, 0x8100
+ * 0x88A8, 0x8100 (outer, inner)
+ * 0x8100, 0x8100 (outer, inner)
  * 0x8100
  * 0x88A8
  */
@@ -2013,8 +2013,8 @@ static int mv_pp2x_prs_vlan_init(struct platform_device *pdev,
 					    GFP_KERNEL);
 	if (!hw->prs_double_vlans)
 		return -ENOMEM;
-	/* Double VLAN: 0x8100, 0x88A8 */
-	err = mv_pp2x_prs_double_vlan_add(hw, ETH_P_8021Q, ETH_P_8021AD,
+	/* Double VLAN: 0x88A8, 0x8100 */
+	err = mv_pp2x_prs_double_vlan_add(hw, ETH_P_8021AD, ETH_P_8021Q,
 					  MVPP2_PRS_PORT_MASK);
 	if (err)
 		return err;
-- 
1.7.9.5

